lib/moab/file_signature.rb

```diff
@@ -65,6 +60,30 @@ def initialize(opts = {})
     # @return [String] The SHA256 checksum value of the file
     attribute :sha256, String, :on_save => Proc.new { |n| n.nil? ? "" : n.to_s }
 
+    KNOWN_ALGOS = {
+      md5: proc { Digest::MD5.new },
+      sha1: proc { Digest::SHA1.new },
+      sha256: proc { Digest::SHA2.new(256) }
+    }.freeze
+
+    # Reads the file once for ALL (requested) algorithms, not once per.
+    # @param [Pathname] pathname
+    # @param [Array<Symbol>] one or more keys of KNOWN_ALGOS to be computed
+    # @return [Moab::FileSignature] object populated with (requested) checksums
+    def self.from_file(pathname, algos_to_use = KNOWN_ALGOS.keys)
+      raise 'Unrecognized algorithm requested' unless algos_to_use.all? { |a| KNOWN_ALGOS.include?(a) }
+
+      signatures = algos_to_use.map { |k| [k, KNOWN_ALGOS[k].call] }.to_h
+
+      pathname.open("r") do |stream|
+        while (buffer = stream.read(8192))
```

Comment:
in a perfect world, would we also want to call out 8192 as a constant and give it the same treatment?  (is it used anywhere else?).   Again, i would suggest creating a ticket here and consider it non-critical path ...

Reply:
Maybe, but this is a feature in search of use case right now.  Until we have a second value we want to use, I wouldn't touch it.

Reply:
IIRC, @tingulfsen did some testing around this and found that it didn't make a difference once you got past a certain size, but i think a config param would be a totally reasonable backlog ticket.

Reply:
#144 created.